src/base/ClpMap.h

```diff
@@ -0,0 +1,266 @@
+
+    static uint64_t MemoryCountedFor(const Key &, const Value &);
+
+    void trim(uint64_t wantSpace);
+    void erase(const IndexIterator &);
+    IndexIterator find(const Key &);
+
+    /// cached entries, including expired ones, in LRU order
+    Entries entries_;
+
+    /// entries_ positions indexed by the entry key
+    Index index_;
+
+    /// entry TTL to use if none provided to add()
+    Ttl defaultTtl_ = std::numeric_limits<Ttl>::max();
+
+    /// the maximum memory we are allowed to use for all cached entries
+    uint64_t memLimit_ = 0;
+
+    /// the total amount of memory we currently use for all cached entries
+    uint64_t memUsed_ = 0;
+};
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+ClpMap<Key, Value, MemoryUsedBy>::ClpMap(const uint64_t capacity, const Ttl defaultTtl):
+    defaultTtl_(defaultTtl)
+{
+    assert(defaultTtl >= 0);
+    setMemLimit(capacity);
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+void
+ClpMap<Key, Value, MemoryUsedBy>::setMemLimit(const uint64_t newLimit)
+{
+    assert(newLimit >= 0);
+    if (memUsed_ > newLimit)
+        trim(memLimit_ - newLimit);
+    memLimit_ = newLimit;
+}
+
+/// \returns the index position of an entry identified by its key (or end())
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+typename ClpMap<Key, Value, MemoryUsedBy>::IndexIterator
+ClpMap<Key, Value, MemoryUsedBy>::find(const Key &key)
+{
+    const auto i = index_.find(key);
+    if (i == index_.end())
+        return i;
+
+    const auto entryPosition = i->second;
+    if (!entryPosition->expired()) {
+        if (entryPosition != entries_.begin())
+            entries_.splice(entries_.begin(), entries_, entryPosition);
+        return i;
+    }
+    // else fall through to cleanup
+
+    erase(i);
+    return index_.end();
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+const Value *
+ClpMap<Key, Value, MemoryUsedBy>::get(const Key &key)
+{
+    const auto i = find(key);
+    if (i != index_.end()) {
+        const auto &entry = *(i->second);
+        return &entry.value;
+    }
+    return nullptr;
+}
+
+template <class Key, class Value, uint64_t MemoryUsedBy(const Value &)>
+uint64_t
+ClpMap<Key, Value, MemoryUsedBy>::MemoryCountedFor(const Key &k, const Value &v)
+{
+    // Both storage and index store keys, but we count keySz once, assuming that
+    // copying a Key does not consume more memory. This assumption holds for
+    // Key=SBuf, but, ideally, we should be outsourcing this decision to another
+    // configurable function, storing each key once, or hard-coding Key=SBuf.
+    const auto keySz = k.length();
+
+    // approximate calculation (e.g., containers store wrappers not value_types)
+    const auto storageSz = sizeof(typename Entries::value_type) + MemoryUsedBy(v);
+    const auto indexSz = sizeof(typename Index::value_type);
+    return keySz + storageSz + indexSz;
```

Comment:
MemoryCountedFor() sums uint64_t values without checking for overflows. I suggest returning zero in overflow cases. The caller should be adjusted to reject the corresponding values. Since we always add a few sizeof()s, the zero return value cannot happen naturally. @yadij, any better ideas? Would you be able to add this code?

FWIW, I doubt we should throw (and try/catch) on overflows because I suspect that doing it correctly would require introducing a special exception value and quite a bit of other custom code.

We could also return uint64::max() on overflows, which feels natural, but that large value, unlike zero, is, technically, a possible/valid value for this function.

Reply:
Agreed on zero.

Reply:
Done in 260e3bbfc